hircluster.c

```diff
@@ -3773,6 +3697,124 @@ static void redisClusterAsyncCallback(redisAsyncContext *ac, void *r,
     cluster_async_data_free(cad);
 }
 
+/* Reply callback function for CLUSTER SLOTS */
+void clusterSlotsReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisReply *reply = (redisReply *)r;
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+
+    if (reply == NULL) {
+        /* Retry using available nodes */
+        if (updateSlotMapAsync(acc) == REDIS_OK) {
+            acc->update_route_time = SLOTMAP_UPDATE_ONGOING;
+        } else {
+            acc->update_route_time =
+                hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+        }
+        return;
+    }
+
+    dict *nodes = parse_cluster_slots(cc, reply, cc->flags);
+    if (updateNodesAndSlotmap(cc, nodes) != REDIS_OK) {
+        /* Ignore failures for now */
+    }
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+}
+
+/* Reply callback function for CLUSTER NODES */
+void clusterNodesReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisReply *reply = (redisReply *)r;
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+
+    if (reply == NULL) {
+        /* Retry using available nodes */
+        if (updateSlotMapAsync(acc) == REDIS_OK) {
+            acc->update_route_time = SLOTMAP_UPDATE_ONGOING;
+        } else {
+            acc->update_route_time =
+                hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+        }
+        return;
+    }
+
+    dict *nodes = parse_cluster_nodes(cc, reply->str, reply->len, cc->flags);
+    if (updateNodesAndSlotmap(cc, nodes) != REDIS_OK) {
+        /* Ignore failures for now */
+    }
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+}
+
+/* Select a node.
+ * Primarily select a connected node by picking one of four first found
+ * connected nodes. If there are no connected nodes then pick a node
+ * where a connect has not been attempted within throttle time (1 sec).
+ */
+static redisClusterNode *selectNode(dict *nodes) {
+    redisClusterNode *node, *selected = NULL;
+    dictIterator di;
+    dictInitIterator(&di, nodes);
+
+    int foundConnected = 0;
+    int64_t throttleLimit = hi_usec_now() - SLOTMAP_UPDATE_THROTTLE_USEC;
+
+    dictEntry *de;
+    while ((de = dictNext(&di)) != NULL) {
+        node = dictGetEntryVal(de);
+        if (node->acon != NULL && node->acon->err == 0 &&
+            node->acon->c.flags & REDIS_CONNECTED) {
+            foundConnected += 1;
+
+            if (foundConnected == 1) {
+                selected = node; /* Always keep the first found */
+            } else if ((random() % 4 + 1) == 4) {
+                selected = node; /* Select node when a random value 1-4 is 4 */
+            }
```

Comment:
Why random() in every iteration in the loop? It's hard to see this is a fair random. Why not pick a random index before the loop?

Reply:
Sorry, I see now why we want to pick the current node with 25% probability. It works regardless of how many nodes we have. (I guess it's not a fair distribution if we have only 2 or 3 nodes, but I'm not sure it's important. Or is it? Maybe 3 is a very common number of masters in a small cluster...)

Reply:
Updated now to first perform 3 attempts to randomly pick a connected node, or else select any node that has not been attempted very recently.

Reply:
Yes, it looks very simple and nice now! I like it.